### PR TITLE
Add multi arch support

### LINKED ISSUE #709: Multi-Arch images
#### Summary

Currently the docker images are amd64-only.
I've tried building the image on an aarch64 machine and it worked out of the box, although MS SQL probably wouldn't work as [these](https://github.com/shlinkio/shlink/blob/develop/Dockerfile#L27) binaries are amd64.

Having multi-arch images would make running shlink on devices like the Raspberry Pi much easier.

An easy way to build multi-arch images is `docker buildx` like here: https://github.com/Starbix/docker-plex-ldap/blob/master/.github/workflows/build.yml

---

https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

### PR DESCRIPTION
**Pull request by Starbix**
Closes #709

### CODE REVIEW

**Comment by acelaya** [comment]
Thanks @Starbix 

Let me know when it's ready to review.

File: docker/build [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
 
+BUILDX_VER=v0.4.1
+DOCKER_CLI_EXPERIMENTAL=enabled
+
+mkdir -vp ~/.docker/cli-plugins/ ~/dockercache
+curl --silent -L "https://github.com/docker/buildx/releases/download/${BUILDX_VER}/buildx-${BUILDX_VER}.linux-amd64" > ~/.docker/cli-plugins/docker-buildx
+chmod a+x ~/.docker/cli-plugins/docker-buildx
+
+docker run --rm --privileged multiarch/qemu-user-static:register --reset
+docker buildx create --use
```
[the remark this answers is not in the document]
Comment (Starbix) [reply docker/build:12]: As far as I understand buildx is a docker plugin, which is only bundled in newer docker releases, which is why it needs to be downloaded at the moment. More info here: https://github.com/docker/buildx#building-with-buildx
multiarch/qemu-user-static:register tells the kernel that programs with an architecture other than the native one can be executed by QEMU. https://github.com/multiarch/qemu-user-static
The code is adapted from here: https://github.com/jdrouet/docker-on-ci/ which I found here: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

File: docker/build [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
 echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
 
 # If there is a tag, regardless the branch, build that docker tag and also "stable"
 if [[ ! -z $TRAVIS_TAG ]]; then
-  docker build --build-arg SHLINK_VERSION=${TRAVIS_TAG#?} -t shlinkio/shlink:${TRAVIS_TAG#?} -t shlinkio/shlink:stable .
-  docker push shlinkio/shlink:${TRAVIS_TAG#?}
-  docker push shlinkio/shlink:stable
+  docker buildx build --push \
+		--build-arg SHLINK_VERSION=${TRAVIS_TAG#?} \
+		--platform linux/arm/v7,linux/arm64/v8,linux/amd64 \
```
[the remark this answers is not in the document]
Comment (Starbix) [reply docker/build:20]: I chose those platforms as they are the most used ones. However the following are possible: `linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64,linux/ppc64le,linux/s390x`

[the remark this answers is not in the document]
Comment (acelaya) [reply docker/build:20]: Ok, sounds good. I guess it shouldn't be hard to change later if needed.

File: docker/build (hunk 1 above)
[the remark this answers is not in the document]
Comment (acelaya) [reply docker/build:12]: Ok, thanks for the detailed explanation.

I also had that article bookmarked, but didn't have the chance to read it yet.

File: Dockerfile [hunk 3]
```diff
@@ -23,18 +23,25 @@ RUN \
...
+      apk del .phpize-deps && \
+      rm msodbcsql17_17.5.1.1-1_amd64.apk && \
+      rm mssql-tools_17.5.1.1-1_amd64.apk ; \
+    fi
 
+# Install swoole
+RUN apk add --no-cache --virtual .phpize-deps $PHPIZE_DEPS && \
+    pecl install swoole-${SWOOLE_VERSION} && \
+    docker-php-ext-enable swoole && \
+    apk del .phpize-deps
```
Comment (Starbix) [inline Dockerfile:45]: I've split up the swoole and mssql installation as the mssql tools are amd64 only.

File: .travis.yml [hunk 4]
```diff
@@ -15,6 +17,11 @@ cache:
     - $HOME/.composer/cache/files
 
 before_install:
+  - curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+  - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+  - sudo apt-get update
+  - sudo apt-get -y -o Dpkg::Options::="--force-confnew" install docker-ce
+  - docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
   - sudo ./data/infra/ci/install-ms-odbc.sh
```
Comment (Starbix) [inline .travis.yml:25]: It now first installs the latest docker version, as 18.06 didn't work for me.

**Comment by acelaya** [comment]
Hey @Starbix.

I see there are a few conflicts in this PR. Ping me here when you fixed them and I'll give the PR a review.

**Comment by Starbix** [comment]
@acelaya I just fixed the conflicts

File: .travis.yml [hunk 5]
```diff
@@ -15,6 +17,11 @@ cache:
     - $HOME/.composer/cache/files
 
 before_install:
+  - curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+  - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+  - sudo apt-get update
+  - sudo apt-get -y -o Dpkg::Options::="--force-confnew" install docker-ce
+  - docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```
Comment (acelaya) [inline .travis.yml:24]: Would you mind putting all these instructions on a shell script?

You can put it on the `data/infra/ci` folder, next to `install-ms-odbc.sh`.

I think it will have a more clear intention if we have a call to a script with a meaningful name.

File: Dockerfile [hunk 6]
```diff
@@ -23,15 +23,22 @@ RUN \
...
+# Install sqlsrv driver
+RUN if [ $(uname -m) == "x86_64" ]; then \
+      wget https://download.microsoft.com/download/e/4/e/e4e67866-dffd-428c-aac7-8d28ddafb39b/msodbcsql17_17.5.1.1-1_amd64.apk && \
+      apk add --allow-untrusted msodbcsql17_17.5.1.1-1_amd64.apk && \
+      apk add --no-cache --virtual .phpize-deps $PHPIZE_DEPS unixodbc-dev && \
+      pecl install pdo_sqlsrv && \
+      docker-php-ext-enable pdo_sqlsrv && \
+      apk del .phpize-deps && \
+      rm msodbcsql17_17.5.1.1-1_amd64.apk ; \
+    fi
```
Comment (acelaya) [inline Dockerfile:35]: Have you checked if there's versions of these binaries for other platforms?

I'm ok with leaving mssql support out for some platforms if there's no support, but just to make sure, as I will have to document it.

Comment (Starbix) [reply Dockerfile:35]: I don't see any arm64 binaries on here unfortunately: https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server

File: .travis.yml (hunk 5 above)
Comment (Starbix) [reply .travis.yml]: Thanks for the review!

File: Dockerfile (hunk 6 above)
Comment (acelaya) [reply Dockerfile:35]: We will have to live with that then. Thanks for checking

**Comment by acelaya** [comment]
Hey @Starbix.

I'm manually trying to verify that nothing will break once this is merged (as there's no way to test it before), and I'm trying to generate and push a dummy image, but I'm getting some errors.

I have locally run the `docker/build` script, by changing the image name (so that it doesn't overwrite real images), and this is what's happening when the `docker buildx build` command is run:

```bash
+ docker buildx build --push --platform linux/arm/v7,linux/arm64/v8,linux/amd64 -t shlinkio/lab:latest .
[+] Building 11.1s (6/21)                                                                                                                                                                                                               
 => [internal] booting buildkit                                                                                                                                                                                                    2.7s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                                 2.1s
 => => creating container buildx_buildkit_quizzical_babbage90                                                                                                                                                                      0.6s
 => [internal] load build definition from Dockerfile                                                                                                                                                                               0.0s
 => => transferring dockerfile: 2.54kB                                                                                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                  0.0s
 => => transferring context: 348B                                                                                                                                                                                                  0.0s
 => [linux/arm/v7 internal] load metadata for docker.io/library/php:7.4.5-alpine3.11                                                                                                                                               8.3s
 => [linux/amd64 internal] load metadata for docker.io/library/php:7.4.5-alpine3.11                                                                                                                                                8.3s
 => [linux/arm64 internal] load metadata for docker.io/library/php:7.4.5-alpine3.11                                                                                                                                                6.8s
failed to solve: rpc error: code = Unknown desc = failed to load LLB: runtime execution on platform linux/arm/v7 not supported
```

**Comment by acelaya** [comment]
I think it's probably related with some syntax error in the `Dockerfile`, because the regular docker build is failing too.

This block has a syntax error:

```
RUN if [ $(uname -m) == "x86_64" ]; then \
      wget https://download.microsoft.com/download/e/4/e/e4e67866-dffd-428c-aac7-8d28ddafb39b/msodbcsql17_17.5.1.1-1_amd64.apk && \
      apk add --allow-untrusted msodbcsql17_17.5.1.1-1_amd64.apk && \
      apk add --no-cache --virtual .phpize-deps $PHPIZE_DEPS unixodbc-dev && \
      pecl install pdo_sqlsrv && \
      docker-php-ext-enable pdo_sqlsrv && \
      apk del .phpize-deps && \
      rm msodbcsql17_17.5.1.1-1_amd64.apk \
    fi
```

**Comment by Starbix** [comment]
My bad, it should be fixed now

**Comment by acelaya** [comment]
The regular build works right now, but running the buildx build still fails for me with the error I shared earlier.

I think it could be because the docker/build script is missing this line

docker run --rm --privileged multiarch/qemu-user-static --reset -p yes

But I can't test it right now. I will continue tomorrow morning.

**Comment by acelaya** [comment]
Hey @Starbix 

I have just fixed the issues I had in my local env and managed to build the multi arch images. You can see the result here https://hub.docker.com/repository/docker/shlinkio/lab/tags?page=1

I have also run the image and it works like a charm (just the amd64 one, but I will test the rest later, on a raspberry).

The only thing that was still failing for me was the `docker buildx create --use` command, which was saying it couldn't generate a random name.

However, since the script is going to be run in freshly spinned-up VMs in travis, I guess it will work. If it does not, at least I know were the issue is.

So, long story short, I'm merging this. Thanks a lot for this huge contribution. This is preciselly the way in which I want Shlink to evolve :slightly_smiling_face:

**Comment by Starbix** [comment]
Thanks a lot for merging multi-arch support and for the amazing software that Shlink is 😄

I'll also open a PR for the Shlink web-client.

**Comment by acelaya** [comment]
That sounds great :slightly_smiling_face:

**Comment by acelaya** [comment]
Hey @Starbix 

Just in case you are wondering why there's still no docker images published for architectures other than amd64, it's because I run in some problems during travis executions with this new approach.

Apparently, building all the platform images with docker buildx is crazy slow in travis, and makes the build hit the time limit (which is set at 50 minutes), which automatically cancels the build.

I tried to work around it by changing the approach in travis. It now runs a job with all the existing tasks, and a parallel job which is only responsible of building and publishing the docker images. It basically runs the new `install-docker.sh` script and then the `docker/build` one.

However, this job is still taking more than 50 minutes on itself.

I also tried splitting the build of every platform on its own parallel job, but then when trying pu push the tag only the last one prevails, overriding other platforms.

For now I have allowed this new job to fail, to not affect the whole build (this is an example https://travis-ci.org/github/shlinkio/shlink/builds/687882992).

I will continue investigating why is this taking so much longer than before, but also some alternative solutions (github actions have crossed my mind), but for now docker image builds are not guaranteed to succeed.

**Comment by Starbix** [comment]
Building the image natively would speed things up. It's possible to add arm64 machines as Github Action Workers. If you have a arm64 SoC laying around, you could try that, or trust a 3rd party running one (I may soon have an unused arm64 board).

**Comment by acelaya** [comment]
Ok, I will investigate on that regards. I think Travis also added support recently for builds on to of arm architectures.

Another thing I tried was splitting the build for every architecture in parallel Travis jobs, but I didn't find a way to push one architecture to docker registry that gets appended to other architectures on an existing tag. Instead I was always getting the whole tag overwritten, leaving just the latest of the architectures to be pushed.

**Comment by acelaya** [comment]
@Starbix it's been a while, but Shlink multi-arch image builds are finally active.

I have moved to github actions to build the docker image. It's still a bit slow, but it has way higher execution time limits, so it should be ok.

https://github.com/shlinkio/shlink/runs/933370133?check_suite_focus=true

The tag `latest` is already published for all architectures, and v2.3.0 will be the first release with support for multi-arch.

Thanks again for the contribution and your patience.

**Comment by Starbix** [comment]
That's great to hear! I haven't been very active on Github the past few weeks, but I'll be able to test shlink on aarch64 in about a week. 😃
